### DISCUSSION TITLE
expose browser window size to python code

### DISCUSSION THREAD

**Discussion by ghomem** [bokeh/bokeh#11471] [General]
### General idea

```
if width > = length:
    # horizontal
    layout1, layout2, layout3, controls1, controls2 = make_layouts(True)
else:
    # vertical
    layout1, layout2, layout3, controls1, controls2 = make_layouts(False)

curdoc().clear()

curdoc().add_root(controls1)
curdoc().add_root(layout1)

curdoc().add_root(controls2)
curdoc().add_root(layout2)

curdoc().add_root(layout3)
```
### Motivation

Concern separation between the python code and the HTML/CSS code. Use of beautiful HTML/CSS templates with central placeholder for bokeh layouts.

### Current situation

The code above works but I did not find a way to access the browser window dimensions apart from a callback triggered by a widget. We are able to recreate the layouts if someone pushes a button but not on window resize and not on the page load.

Example:

```
# this callback runs on the browser and updates the server side data source from the dimensions data got via Javascript
mycallback = CustomJS( args=dict(ds=window_size_data_source), code="""
var width, height;
var new_data = {};
height = window.innerHeight|| document.documentElement.clientHeight|| document.body.clientHeight;
width  = window.innerWidth || document.documentElement.clientWidth || document.body.clientWidth;
console.log("Javascript callback", height, width);

// for some reason the right side needs to be a list, otherwise we have a server side error
new_data['height'] = [height];
new_data['width' ] = [width];
ds.data = new_data;
ds.change.emit();

""")

# this callbacks takes action on the server side upon dimensions change
def on_source_change(attr, old, new):

    width  = new['width'][0]
    height = new['height'][0]

    print('current dimensions', width, height)

    if width >= height:
        print('orientation is horizontal')
        # switch layout to verticals python side

    else:
        print('orientation is vertical')
        # switch layout to horizontal python side


# this connects a data change on the data source to the server side callback
window_size_data_source.on_change('data', on_source_change)

# TODO: how to have this on a window resize event and during the page load
date_slider_map.js_on_change('value_throttled', mycallback)
```

### To be discussed

Would it be possible to expose the window dimensions (height, width) to the python code? And to run callbaks on page load and page resize?

**Comment by ghomem** [comment]
@mattpap @philippjfr pinging as suggested by Bryan. Thanks!

**Comment by bryevdv** [comment]
There are probably various ways to attack this. E.g. once there is support for a `CustomJS` hook for "document init" then a callback there could set a property on a simple `DataClass` to trigger some Python event (or perhaps the JS callback could just update the layout itself, even easier)

But this is at least tangential to some other questions:

* Would it make sense to have a `PageMetrics` model that we automatically update? 
* Should there be a notion of "document-level" objects like a page metrics or a data-class "scratch variable" space?
* Should we make it easy for users to define and emit custom events?

**Reply by ghomem** [reply]
Thanks for paying attention to this.

My personal preference would be centralizing the layout work in the python code, because of centralization of concerns. If it is possible to have the document init event trigger python code, that would be brilliant.

But having a PageMetrics model that you automatically update would be even better. If we could hookup a python callback to a PageMetrics instance by means of on_change and have it run automatically each time the page changes the solution would be clean and python centric.

```
page_metrics = curdoc().get_page_metrics()
page_metrics.on_change('metrics_throttled', mycallback) # metrics vs metrics_trottled?  like have for sliders perhaps
```
Which events would trigger the on_change then? The obvious one is page resize.

Ideally, the PageMetrics model would know if the page ratio is horizontal or vertical at each moment, so that we don't have to carry an "if" all the time :-)

**Comment by mattpap** [comment]
We can definitively expose viewport's properties for special use cases, but for responsive UI this needs to be handled in browser, otherwise the lag will be unbearable (imagine user flipping a phone and having to wait for a server to respond). Given that will be overhauling bokeh's layout for 3.0 and exposing CSS layout more directly, we will add support for proper responsive layouts (not just resizing to viewport size).

**Reply by ghomem** [reply]
> We can definitively expose viewport's properties for special use cases, but for responsive UI this needs to be handled in browser, otherwise the lag will be unbearable (imagine user flipping a phone and having to wait for a server to respond). Given that will be overhauling bokeh's layout for 3.0 and exposing CSS layout more directly, we will add support for proper responsive layouts (not just resizing to viewport size).

I already tried this code
```
f width > = length:
    # horizontal
    layout1, layout2, layout3, controls1, controls2 = make_layouts(True)
else:
    # vertical
    layout1, layout2, layout3, controls1, controls2 = make_layouts(False

curdoc().clear()
curdoc().add_root(controls1)
curdoc().add_root(layout1)
curdoc().add_root(controls2)
curdoc().add_root(layout2)
curdoc().add_root(layout3)
```
and it runs quite fast because it does not involve any recalculation of the plots (if it did it would be really slow, but it is not the case here).

**Reply by ghomem** [reply]
>  we will add support for proper responsive layouts (not just resizing to viewport size

this great as well :-)
